src/util/settings_persister.js

```diff
@@ -21,6 +21,28 @@ export const localStorageAvailable = (() => {
   }
 })();
 
+const updateConfigForGitLab = async (client, contents) => {
+  const filename = '/.organice-config.json';
+  // GitLab doesn't allow updating a file that doesn't exist or creating one that already exists, so
+  // need to figure out whether or not it exists.
+  let exists = false;
+  try {
+    const existingContents = await client.getFileContents(filename);
+    if (existingContents === contents) {
+      // Also avoid a pointless empty commit if file didn't actually change.
+      return;
+    }
```

Comment:
If you're sure that redundant updates were occurring without this additional check, then I guess adding version control has yielded the first bug in the legacy code(; As you said, `subscribeToChanges` should have avoided this already, but it didn't.

If you do find a fix for the original code, that would benefit all sync back-ends. If you don't then this is certainly a good workaround for a Git based back-end. Doing additional pushes on the other ones aren't optimal, but shouldn't cause a lot of trouble.

Reply:
Okay, so I haven't fixed it because the issue is non-trivial, but I more or less figured out what's going on. Turns out there are two different issues causing redundant config syncs:

1. `subscribeToChanges` does its comparison outside of the debounce logic, which means that an extraneous sync will be triggered if you change a setting and then quickly change it back to the previous value since it does not take the debounce into consideration. I think the most straightforward way to solve this would be to move the comparison and `window.previousSettingsFileContents` assignment into `debouncedPushConfigToSyncBackend`. See  this section of `subscribeToChanges`: https://github.com/200ok-ch/organice/blob/fbdcf5f040218373b2b62811fbe9648ebdb6efc0/src/util/settings_persister.js#L392-L398
2. The settings initially loaded from `localStorage` include default values, whereas later instances of the settings object do not. So the first time `subscribeToChanges` is called will always result in a sync due to a different number of fields, unless you've changed every setting to avoid defaults being relevant. So either the non-initial settings objects need to have defaults added, or the initial object needs to avoid adding defaults, not sure which is the right solution at the moment. I would suggest whoever works on this start by looking at occurences of `.default` in `settings_persister.js` as well entry point when loading initial settings from `localStorage`, which is here: https://github.com/200ok-ch/organice/blob/fbdcf5f040218373b2b62811fbe9648ebdb6efc0/src/util/settings_persister.js#L321-L325

---

So there are definitely bug fixes to be made in `settings_persister.js`, but I think it's way out of scope for this PR. How do you feel about merging this as-is and making a new issue to fix these two bugs? I can *probably* work on a PR for that too, but I'm not positive when or if I'll get around to it.

Reply:
Thank you for digging into this further! That work definitively brought the task into a state so that anyone could take it up at some point. Hence, I have created an issue https://github.com/200ok-ch/organice/issues/736

It's definitively not a blocker for this PR.